In [18]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression


In [3]:
credit_df=pd.read_csv('Credit_card.csv')
credit_df.head()


Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
0  5008827      M         Y             Y         0       180000.0   
1  5009744      F         Y             N         0       315000.0   
2  5009746      F         Y             N         0       315000.0   
3  5009749      F         Y             N         0            NaN   
4  5009752      F         Y             N         0       315000.0   

            Type_Income         EDUCATION Marital_status       Housing_type  \
0             Pensioner  Higher education        Married  House / apartment   
1  Commercial associate  Higher education        Married  House / apartment   
2  Commercial associate  Higher education        Married  House / apartment   
3  Commercial associate  Higher education        Married  House / apartment   
4  Commercial associate  Higher education        Married  House / apartment   

   Birthday_count  Employed_days  Mobile_phone  Work_Phone  Phone  EMAIL_ID  \
0        -18772.0         365243             1           0      0         0   
1        -13557.0           -586             1           1      1         0   
2             NaN           -586             1           1      1         0   
3        -13557.0           -586             1           1      1         0   
4        -13557.0           -586             1           1      1         0   

  Type_Occupation  Family_Members  
0             NaN               2  
1             NaN               2  
2             NaN               2  
3             NaN               2  
4             NaN               2

In [4]:
credit_df.dtypes

Ind_ID               int64
GENDER                 str
Car_Owner              str
Propert_Owner          str
CHILDREN             int64
Annual_income      float64
Type_Income            str
EDUCATION              str
Marital_status         str
Housing_type           str
Birthday_count     float64
Employed_days        int64
Mobile_phone         int64
Work_Phone           int64
Phone                int64
EMAIL_ID             int64
Type_Occupation        str
Family_Members       int64
dtype: object

In [5]:
# 1. Data cleaning
credit_df = credit_df.drop(columns=['Ind_ID','Mobile_phone','Work_Phone','Phone','EMAIL_ID'])

In [6]:
# Convert categorical variables to numeric
credit_df = pd.get_dummies(credit_df, columns=['GENDER', 'Car_Owner', 'Propert_Owner', 'Type_Income','EDUCATION', 'Marital_status','Housing_type','Type_Occupation'], drop_first=True)
credit_df.head()

CHILDREN  Annual_income  Birthday_count  Employed_days  Family_Members  \
0         0       180000.0        -18772.0         365243               2   
1         0       315000.0        -13557.0           -586               2   
2         0       315000.0             NaN           -586               2   
3         0            NaN        -13557.0           -586               2   
4         0       315000.0        -13557.0           -586               2   

   GENDER_M  Car_Owner_Y  Propert_Owner_Y  Type_Income_Pensioner  \
0      True         True             True                   True   
1     False         True            False                  False   
2     False         True            False                  False   
3     False         True            False                  False   
4     False         True            False                  False   

   Type_Income_State servant  ...  Type_Occupation_Laborers  \
0                      False  ...                     False   
1                      False  ...                     False   
2                      False  ...                     False   
3                      False  ...                     False   
4                      False  ...                     False   

   Type_Occupation_Low-skill Laborers  Type_Occupation_Managers  \
0                               False                     False   
1                               False                     False   
2                               False                     False   
3                               False                     False   
4                               False                     False   

   Type_Occupation_Medicine staff  Type_Occupation_Private service staff  \
0                           False                                  False   
1                           False                                  False   
2                           False                                  False   
3                           False                                  False   
4                           False                                  False   

   Type_Occupation_Realty agents  Type_Occupation_Sales staff  \
0                          False                        False   
1                          False                        False   
2                          False                        False   
3                          False                        False   
4                          False                        False   

   Type_Occupation_Secretaries  Type_Occupation_Security staff  \
0                        False                           False   
1                        False                           False   
2                        False                           False   
3                        False                           False   
4                        False                           False   

   Type_Occupation_Waiters/barmen staff  
0                                 False  
1                                 False  
2                                 False  
3                                 False  
4                                 False  

[5 rows x 41 columns]

In [7]:
# replace NAN with the median
credit_df['Birthday_count']=credit_df['Birthday_count'].fillna(credit_df['Birthday_count'].median())
credit_df['Annual_income']=credit_df['Annual_income'].fillna(credit_df['Annual_income'].median())
credit_df['CHILDREN']=credit_df['CHILDREN'].fillna(credit_df['CHILDREN'].median())


In [8]:
credit_df.head()

CHILDREN  Annual_income  Birthday_count  Employed_days  Family_Members  \
0         0       180000.0        -18772.0         365243               2   
1         0       315000.0        -13557.0           -586               2   
2         0       315000.0        -15661.5           -586               2   
3         0       166500.0        -13557.0           -586               2   
4         0       315000.0        -13557.0           -586               2   

   GENDER_M  Car_Owner_Y  Propert_Owner_Y  Type_Income_Pensioner  \
0      True         True             True                   True   
1     False         True            False                  False   
2     False         True            False                  False   
3     False         True            False                  False   
4     False         True            False                  False   

   Type_Income_State servant  ...  Type_Occupation_Laborers  \
0                      False  ...                     False   
1                      False  ...                     False   
2                      False  ...                     False   
3                      False  ...                     False   
4                      False  ...                     False   

   Type_Occupation_Low-skill Laborers  Type_Occupation_Managers  \
0                               False                     False   
1                               False                     False   
2                               False                     False   
3                               False                     False   
4                               False                     False   

   Type_Occupation_Medicine staff  Type_Occupation_Private service staff  \
0                           False                                  False   
1                           False                                  False   
2                           False                                  False   
3                           False                                  False   
4                           False                                  False   

   Type_Occupation_Realty agents  Type_Occupation_Sales staff  \
0                          False                        False   
1                          False                        False   
2                          False                        False   
3                          False                        False   
4                          False                        False   

   Type_Occupation_Secretaries  Type_Occupation_Security staff  \
0                        False                           False   
1                        False                           False   
2                        False                           False   
3                        False                           False   
4                        False                           False   

   Type_Occupation_Waiters/barmen staff  
0                                 False  
1                                 False  
2                                 False  
3                                 False  
4                                 False  

[5 rows x 41 columns]

In [9]:
correlation = credit_df.corr()
correlation

CHILDREN  Annual_income  \
CHILDREN                                 1.000000       0.078543   
Annual_income                            0.078543       1.000000   
Birthday_count                           0.277944       0.110682   
Employed_days                           -0.219095      -0.158966   
Family_Members                           0.890248       0.051522   
GENDER_M                                 0.063068       0.207970   
Car_Owner_Y                              0.063467       0.208948   
Propert_Owner_Y                         -0.003352       0.043012   
Type_Income_Pensioner                   -0.217467      -0.145588   
Type_Income_State servant                0.035128       0.051633   
Type_Income_Working                      0.095880      -0.093423   
EDUCATION_Higher education               0.052443       0.246427   
EDUCATION_Incomplete higher              0.036255       0.006856   
EDUCATION_Lower secondary               -0.012001      -0.055724   
EDUCATION_Secondary / secondary special -0.064543      -0.223767   
Marital_status_Married                   0.145940      -0.021300   
Marital_status_Separated                 0.001284      -0.001372   
Marital_status_Single / not married     -0.133364       0.065766   
Marital_status_Widow                    -0.096715      -0.065196   
Housing_type_House / apartment           0.025007       0.009001   
Housing_type_Municipal apartment        -0.045196      -0.039948   
Housing_type_Office apartment            0.003118       0.045559   
Housing_type_Rented apartment           -0.012001       0.021355   
Housing_type_With parents                0.014946       0.003787   
Type_Occupation_Cleaning staff          -0.007601      -0.069350   
Type_Occupation_Cooking staff           -0.026385      -0.041685   
Type_Occupation_Core staff               0.050506      -0.006547   
Type_Occupation_Drivers                  0.070828       0.081366   
Type_Occupation_HR staff                 0.033326      -0.016671   
Type_Occupation_High skill tech staff    0.013144       0.031807   
Type_Occupation_IT staff                -0.019122      -0.028006   
Type_Occupation_Laborers                 0.022808      -0.023870   
Type_Occupation_Low-skill Laborers      -0.007826      -0.040497   
Type_Occupation_Managers                 0.052493       0.274205   
Type_Occupation_Medicine staff           0.029932      -0.052358   
Type_Occupation_Private service staff    0.031794       0.008767   
Type_Occupation_Realty agents            0.027202       0.003670   
Type_Occupation_Sales staff              0.017414      -0.030810   
Type_Occupation_Secretaries              0.014061      -0.019408   
Type_Occupation_Security staff          -0.002111      -0.010724   
Type_Occupation_Waiters/barmen staff    -0.015601      -0.034754   

                                         Birthday_count  Employed_days  \
CHILDREN                                       0.277944      -0.219095   
Annual_income                                  0.110682      -0.158966   
Birthday_count                                 1.000000      -0.614271   
Employed_days                                 -0.614271       1.000000   
Family_Members                                 0.265066      -0.238705   
GENDER_M                                       0.182646      -0.176383   
Car_Owner_Y                                    0.144234      -0.150658   
Propert_Owner_Y                               -0.124409       0.100686   
Type_Income_Pensioner                         -0.608547       0.981839   
Type_Income_State servant                      0.060405      -0.131034   
Type_Income_Working                            0.278216      -0.464229   
EDUCATION_Higher education                     0.142704      -0.110300   
EDUCATION_Incomplete higher                    0.158108      -0.070678   
EDUCATION_Lower secondary                     -0.016547       0.052027   
EDUCATION_Secondary / secondary special       -0.200437       0.123682   


In [10]:
# 2. Univariate Linear Regression to predict Birthday_count from Employed_days
credit_df_1= credit_df[['Employed_days','Birthday_count']]


X = np.array(credit_df_1['Employed_days']).reshape(-1,1)

y = np.array(credit_df_1['Birthday_count']).reshape(-1,1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42) 

regression_model = LinearRegression()

regression_model.fit(X_train, y_train)

y_pred = regression_model.predict(X_test)

accuracy_score = regression_model.score(X_test, y_test)
print(accuracy_score)


0.3910979159137805


2. Our model did not perform all that well with an accuracy of about 39.1%. Using linear regression is probably not a good choice for this data set as we have lots of discrete data and only really 3 continous variables

In [ ]:
# 3. function that selects the optimal k value
def optimal_k(X_train, y_train, X_test, y_test):
    k_values = range(1, 50)
    best_k = 1
    best_score = 0
    
    for k in k_values:
        knn = KNeighborsClassifier(n_neighbors=k)
        knn.fit(X_train, y_train)
        score = knn.score(X_test, y_test)
        
        if score > best_score:
            best_score = score
            best_k = k
            
    return best_k, best_score

   


In [16]:
# 3. Using KNN to predict if someone is a property owner
X = credit_df.drop('Propert_Owner_Y', axis=1)
y = credit_df['Propert_Owner_Y']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

knn = KNeighborsClassifier(n_neighbors=optimal_k(X_train, y_train, X_test, y_test)[0])

knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)

print(knn.score(X_test, y_test))

0.6344086021505376


Are model performed much better with with an accuracy of about 63.4% which is much better but still not that great

In [22]:
# 4. Logistic Regression
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,
                                                    random_state=42,
                                                    stratify=y)
regression = LogisticRegression(random_state=42, max_iter=10000).fit(X_train, y_train)
y_predicted = regression.predict(X_test)
regression.score(X_test, y_test)

0.6903225806451613

This model performed the best of all of our models with an acurracy of about 69.0% which is better but still not super accurate.

In [23]:
# 5. Linear regression with normalization
credit_df_1= credit_df[['Employed_days','Birthday_count']]


X = np.array(credit_df_1['Employed_days']).reshape(-1,1)

y = np.array(credit_df_1['Birthday_count']).reshape(-1,1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42) 

sc = StandardScaler()

X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

regression_model = LinearRegression()

regression_model.fit(X_train, y_train)

y_pred = regression_model.predict(X_test)

accuracy_score = regression_model.score(X_test, y_test)
print(accuracy_score)

0.3912042876023383


In [24]:
#5. KNN with normalization
X = credit_df.drop('Propert_Owner_Y', axis=1)
y = credit_df['Propert_Owner_Y']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

sc = StandardScaler()

X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

knn = KNeighborsClassifier(n_neighbors=optimal_k(X_train, y_train, X_test, y_test)[0])

knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)

print(knn.score(X_test, y_test))

0.6731182795698925


In [29]:
# 5. Logistic regression with normalization
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,
                                                    random_state=42,
                                                  stratify=y)  

sc = StandardScaler()

X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

regression = LogisticRegression(random_state=42, max_iter=10000).fit(X_train, y_train)
y_predicted = regression.predict(X_test)
regression.score(X_test, y_test)


0.6817204301075269

When we normalized the data, it didn't change the accuracy with linear regression much at all, it improve the KNN accuracy and actually decreased the logisitic regression accuracy. Here it would probably be best to use accuracy since we care more about how correct the model is and not neccesarily about false postives or false negatives